File: api/app/controllers/api/v1/device_settings_controller.rb
# デバイス設定 API（取得/更新）
module Api
  module V1
    # デバイス設定の取得・更新用コントローラ
    class DeviceSettingsController < Api::V1::BaseController
      # ログイン確認
      # before_action :require_login

      # レコード未検出時に 404 を返す想定
      rescue_from ActiveRecord::RecordNotFound do
        render json: { error: "not_found" }, status: :not_found
      end

      # 楽観ロック競合などの更新衝突時に 409 を返す想定
      rescue_from ActiveRecord::StaleObjectError do
        device = Device.find(params[:device_id])
        current = device.device_setting
        render json: { error: "conflict", current_version: current&.lock_version }, status: :conflict
      end

      # バリデーションNGは 422 にする
      rescue_from ActiveRecord::RecordInvalid do |e|
        render json: { error: "unprocessable_entity", messages: e.record.errors.full_messages }, status: :unprocessable_entity
      end

      # GET /api/v1/device_settings/:device_id
      # 指定デバイスの設定を返す。存在しなければ作成して返す。
      def show
        device = Device.find(params[:device_id])
        setting = device.device_setting
        return render json: { error: "not_found" }, status: :not_found unless setting
        render json: serialize(setting)
      end

      # PUT /api/v1/device_settings/:device_id
      def update
        device = Device.find(params[:device_id])
        @setting = device.device_setting
        return render json: { error: "not_found" }, status: :not_found unless @setting
        @setting.update!(device_setting_params)
        render json: serialize(@setting)
      end

      private

      # Strong Parameters: 受け付けるキーを限定
      def device_setting_params
        params.require(:device_setting).permit(
          :stable_duration_sec,
          :max_session_sec,
          :lock_version,
          :tare_weight,
          :stability_epsilon_g,
          :sampling_hz,
          :moving_avg_window,
          :gross_weight_limit_g
          )
      end

      def serialize(s)
        # レスポンスとして返す項目だけを整形
        {
          device_id: s.device_id,
          stable_duration_sec: s.stable_duration_sec,
          max_session_sec: s.max_session_sec,
          lock_version: s.lock_version,
          tare_weight: s.tare_weight,
          stability_epsilon_g: s.stability_epsilon_g,
          sampling_hz: s.sampling_hz,
          moving_avg_window: s.moving_avg_window,
          gross_weight_limit_g: s.gross_weight_limit_g,
          updated_at: s.updated_at
        }
      end
    end
  end
end

File: api/app/Http/Controllers/Api/V1/DeviceSettingsController.php
<?php

namespace App\Http\Controllers\Api\V1;

use App\Http\Controllers\Controller;
use Carbon\CarbonImmutable;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\DB;

class DeviceSettingsController extends Controller
{
    private const REQUIRED_FIELDS = [
        'stable_duration_sec' => 'Stable duration sec',
        'max_session_sec' => 'Max session sec',
        'tare_weight' => 'Tare weight',
        'stability_epsilon_g' => 'Stability epsilon g',
        'sampling_hz' => 'Sampling hz',
        'moving_avg_window' => 'Moving avg window',
        'gross_weight_limit_g' => 'Gross weight limit g',
    ];

    public function show(string $device_id): JsonResponse
    {
        $deviceId = $device_id;
        if (!$this->deviceExists($deviceId)) {
            return $this->notFound();
        }

        $setting = DB::table('device_settings')->where('device_id', $deviceId)->first();
        if (!$setting) {
            return $this->notFound();
        }

        return response()->json($this->serialize($setting));
    }

    public function update(Request $request, string $device_id): JsonResponse
    {
        $deviceId = $device_id;
        $payload = $request->input('device_setting');
        if (!is_array($payload)) {
            return $this->unprocessable(['Device setting is invalid']);
        }

        $errors = $this->validatePayload($payload);
        if (!empty($errors)) {
            return $this->unprocessable($errors);
        }

        if (!$this->deviceExists($deviceId)) {
            return $this->notFound();
        }

        $setting = DB::table('device_settings')->where('device_id', $deviceId)->first();
        if (!$setting) {
            return $this->notFound();
        }

        $lockVersion = $this->parseInteger($payload['lock_version'] ?? null);
        if ($lockVersion === null || $lockVersion < 0) {
            return $this->unprocessable(['Lock version is invalid']);
        }

        $updatedAt = CarbonImmutable::now('UTC');
        $updatedAtValue = $updatedAt->format('Y-m-d H:i:s.u');
        $updateData = [
            'stable_duration_sec' => (int) $payload['stable_duration_sec'],
            'max_session_sec' => (int) $payload['max_session_sec'],
            'tare_weight' => (int) $payload['tare_weight'],
            'stability_epsilon_g' => (int) $payload['stability_epsilon_g'],
            'sampling_hz' => (int) $payload['sampling_hz'],
            'moving_avg_window' => (int) $payload['moving_avg_window'],
            'gross_weight_limit_g' => (int) $payload['gross_weight_limit_g'],
            'lock_version' => $lockVersion + 1,
            'updated_at' => $updatedAtValue,
        ];

        $updated = DB::table('device_settings')
            ->where('device_id', $deviceId)
            ->where('lock_version', $lockVersion)
            ->update($updateData);

        if ($updated === 0) {
            $currentVersion = DB::table('device_settings')
                ->where('device_id', $deviceId)
                ->value('lock_version');

            if ($currentVersion === null) {
                return $this->notFound();
            }

            return response()->json([
                'error' => 'conflict',
                'current_version' => (int) $currentVersion,
            ], 409);
        }

        $setting = (object) array_merge([
            'device_id' => $deviceId,
            'updated_at' => $updatedAtValue,
        ], $updateData);

        return response()->json($this->serialize($setting));
    }

    private function deviceExists(string $deviceId): bool
    {
        return DB::table('devices')->where('id', $deviceId)->exists();
    }

    private function notFound(): JsonResponse
    {
        return response()->json([
            'error' => 'not_found',
        ], 404);
    }

    private function unprocessable(array $messages): JsonResponse
    {
        return response()->json([
            'error' => 'unprocessable_entity',
            'messages' => array_values($messages),
        ], 422);
    }

    private function serialize(object $setting): array
    {
        $updatedAt = $this->formatUtcMillis($setting->updated_at ?? null);

        return [
            'device_id' => (string) $setting->device_id,
            'stable_duration_sec' => (int) $setting->stable_duration_sec,
            'max_session_sec' => (int) $setting->max_session_sec,
            'lock_version' => (int) $setting->lock_version,
            'tare_weight' => (int) $setting->tare_weight,
            'stability_epsilon_g' => (int) $setting->stability_epsilon_g,
            'sampling_hz' => (int) $setting->sampling_hz,
            'moving_avg_window' => (int) $setting->moving_avg_window,
            'gross_weight_limit_g' => (int) $setting->gross_weight_limit_g,
            'updated_at' => $updatedAt,
        ];
    }

    private function formatUtcMillis(?string $value): ?string
    {
        if ($value === null || $value === '') {
            return null;
        }

        $time = CarbonImmutable::parse($value, 'UTC')->setTimezone('UTC');

        return $time->format('Y-m-d\\TH:i:s.v\\Z');
    }

    private function validatePayload(array $payload): array
    {
        $errors = [];

        foreach (self::REQUIRED_FIELDS as $field => $label) {
            if (!array_key_exists($field, $payload) || $payload[$field] === null || $payload[$field] === '') {
                $errors[] = $label . " can't be blank";
                $errors[] = $label . ' is not a number';
                continue;
            }

            $value = $this->parseInteger($payload[$field]);
            if ($value === null) {
                $errors[] = $label . ' is not a number';
                continue;
            }

            if ($value <= 0) {
                $errors[] = $label . ' must be greater than 0';
            }
        }

        return $errors;
    }

    private function parseInteger(mixed $value): ?int
    {
        if (is_int($value)) {
            return $value;
        }

        if (is_string($value) && preg_match('/\A-?\d+\z/', $value) === 1) {
            return (int) $value;
        }

        return null;
    }
}
